pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go

```diff
@@ -0,0 +1,202 @@
+	cmd.Flags().StringVar(&flags.KubeFeatureGates, "kube-feature-gates", vars.KubeFeatureGates, "a set of key=value pairs that describe feature gates for alpha/experimental features of Kubernetes")
+	cmd.Flags().StringVar(&flags.KubeRuntimeConfig, "kube-runtime-config", vars.KubeRuntimeConfig, "a set of key=value pairs that enable or disable built-in APIs")
+	cmd.Flags().DurationVar(&flags.Timeout, "timeout", 30*time.Second, "timeout for waiting for the cluster to be ready")
+	return cmd
+}
+
+func runE(ctx context.Context, logger logger.Logger, flags *flagpole) error {
+	name := fmt.Sprintf("%s-%s", vars.ProjectName, flags.Name)
+	workdir := utils.PathJoin(vars.ClustersDir, flags.Name)
+
+	newRuntime, ok := runtime.Get(flags.Runtime)
+	if !ok {
+		return fmt.Errorf("runtime %q not found", flags.Runtime)
+	}
+
+	dc, err := newRuntime(name, workdir, logger)
+	if err != nil {
+		return err
+	}
+	conf, err := dc.Config()
+	if err == nil {
+		logger.Printf("Cluster %q already exists", name)
+
+		ready, err := dc.Ready(ctx)
+		if err == nil && ready {
+			logger.Printf("Cluster %q is already ready", name)
+			return nil
+		}
+
+		logger.Printf("Cluster %q is not ready yet, will be restarted", name)
+		err = dc.Install(ctx, *conf)
+		if err != nil {
+			logger.Printf("Failed to continue install cluster %q: %v", name, err)
+			return err
+		}
+
+		err = dc.Down(ctx)
+		if err != nil {
+			logger.Printf("Failed to down cluster %q: %v", name, err)
+		}
+	} else {
+		logger.Printf("Creating cluster %q", name)
+		err = dc.Install(ctx, runtime.Config{
+			Name:                        name,
+			ApiserverPort:               flags.ApiserverPort,
+			Workdir:                     workdir,
+			Runtime:                     flags.Runtime,
+			PrometheusImage:             flags.PrometheusImage,
+			EtcdImage:                   flags.EtcdImage,
+			KubeApiserverImage:          flags.KubeApiserverImage,
+			KubeControllerManagerImage:  flags.KubeControllerManagerImage,
+			KubeSchedulerImage:          flags.KubeSchedulerImage,
+			KwokControllerImage:         flags.KwokControllerImage,
+			KindNodeImage:               flags.KindNodeImage,
+			KubeApiserverBinary:         flags.KubeApiserverBinary,
+			KubeControllerManagerBinary: flags.KubeControllerManagerBinary,
+			KubeSchedulerBinary:         flags.KubeSchedulerBinary,
+			KwokControllerBinary:        flags.KwokControllerBinary,
+			EtcdBinaryTar:               flags.EtcdBinaryTar,
+			PrometheusBinaryTar:         flags.PrometheusBinaryTar,
+			CacheDir:                    vars.CacheDir,
+			SecretPort:                  flags.SecurePort,
+			QuietPull:                   flags.QuietPull,
+			PrometheusPort:              flags.PrometheusPort,
+			FeatureGates:                flags.KubeFeatureGates,
+			RuntimeConfig:               flags.KubeRuntimeConfig,
+		})
+		if err != nil {
+			return fmt.Errorf("failed install cluster %q: %w", name, err)
```

Comment:
💭 to brainstorm: should we cleanup the leftover once an error happens? How do other tools (like kind) deal with this situation?

Reply:
Kind no cleanup after a failed cluster creation, to save the scene for troubleshooting.

Reply:
My $0.02 is that it makes more sense to default to "cleanup upon faulty install". Take my UX for example, yesterday I installed it successfully, and it has a "cluster leftover". And today when I retry (with another sets of parameters), it seems to pickup the old config and failed anyway. It takes me some time to figure out the "leftover config was read" and I have to manually delete it.

cc @yibozhuang for thoughts.

Reply:
Personally I would prefer to have the clean up done on a failed install but keep the log around for users to check.
Additionally, having a switch/flag to allow the user to run the install again without clean up for further debug.

We should be as idempotent as possible meaning that a previously failed case should ideally not affect future retries.

Reply:
I added logic that will clean up upon faulty install.